hal/mbed_critical_section_api.c

```diff
@@ -34,17 +35,25 @@ static bool are_interrupts_enabled(void)
 
 MBED_WEAK void hal_critical_section_enter(void)
 {
-    critical_interrupts_enabled = are_interrupts_enabled();
+    const bool interrupt_state = are_interrupts_enabled();
```

Comment:
what if there is a context switch after this line and the actual `PRIMASK` state changes until you reach line 40 (`__disable_irq();`)

Reply:
I imagine it will do the same thing it did before the changes. Adding a solution to get exclusive accesses on boards that don't support it seems like it's going to be a much bigger change than the intent of this PR. It might be best to raise it separately as this PR doesn't introduce the problem, it just highlights it.

Reply:
IMHO `core_util_critical_section_enter` is a public API whichdeclared in [mbed_critical.h#L70](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_critical.h#L70).
This API is widely used within Mbed-OS itself - [find usages here](https://github.com/ARMmbed/mbed-os/search?utf8=%E2%9C%93&q=core_util_critical_section_enter&type=). 

`core_util_critical_section_enter` API directly calls `hal_critical_section_enter` which has problematic implementation. 

We have problem implementing it fully thread safe for ARMv6-M. But for the rest of the platforms we can do proper implementation using exclusive access.